File: template/unipi-values.template.c
/******************************************************************
 *
 *  {{ warning }}
 *
 * uniee_values.c
 *
 *  Created on: Jan 14, 2022
 *      Author: Miroslav Ondra <ondra@faster.cz>
 * 
 */
#include <strings.h>
#include <stddef.h>
#include "uniee.h"

struct slot_dt {
	int slot;
	const char *dt;
};

struct id_name {
	int id;
	const char *name;
	const char *dt;
	const struct slot_dt *slots;
};


struct id_slot_name {
	int id;
	int slot;
	const char *name;
	const char *dt;
};

/* board + slot --> overlays */
{%- for f,x in board_dt.items() %}
const struct slot_dt c_slot_{{f}}[] = {
  {%- for s,v in x.items() %}
  { .slot={{s}}, .dt="{{ v }}" },
  {%- endfor %}
  {/* sentinel */}
};
{%- endfor %}

/* Family names*/
const struct id_name family_table[] = {
{%- for f,v in platform.family.items() %}
  { .id={{ "0x%02x" % v }}, .name="{{ f }}" },
{%- endfor %}
  {/* sentinel */}
};

/* product --> name, overlays */
const struct id_name product_table[] = {
{%- for f,v in platform.id.items() %}
  { .id={{ "0x%04x" % v }}, .name="{{ f }}"
    {%- if v|trim in product_dt -%}
    , .dt="{{ product_dt[v|trim] }}"
    {%- endif -%}
  },
{%- endfor %}
  {/* sentinel */}
};

/* board --> name, overlays, slots */
const struct id_name board_table[] = {
{%- for f,v in board.model.items() %}
  { .id={{ "0x%04x" % v }}, .name="{{ f }}"
    {%- if v|trim in board_dt -%}
    , .slots=c_slot_{{v}}
    {%- endif -%}
    {#- %- if v|trim in board_dt -%}
    , .dt="{{ board_dt[v|trim] }}"
    {%- endif -%  -#}
  },
{%- endfor %}
  {/* sentinel */}
};

static const struct id_name* get_by_name(char *name, const struct id_name *table)
{
	const struct id_name *p;
	for (p=table; p->name != NULL; p++) {
		if (strcasecmp(name, p->name) == 0) return p;
	}
	return NULL;
}

static const struct id_name* get_by_id(int id, const struct id_name *table)
{
	const struct id_name *p;
	for (p=table; p->name != NULL; p++) {
		if (p->id == id) return p;
	}
	return NULL;
}

static const struct slot_dt* get_by_slot(int slot, const struct slot_dt *table)
{
	const struct slot_dt *p;
	for (p=table; p->dt != NULL; p++) {
		if (p->slot == slot) return p;
	}
	return NULL;
}

int unipi_family_by_name(char *name)
{
	const struct id_name *p = get_by_name(name, family_table);
	return p ? p->id : -1;
}

const char* unipi_family_name(platform_id_t id)
{
	const struct id_name *p = get_by_id(id.parsed.platform_family, family_table);
	return p ? p->name : NULL;
}

int unipi_product_by_name(char *name)
{
	const struct id_name *p = get_by_name(name, product_table);
	return p ? p->id : -1;
}

const char* unipi_product_name(platform_id_t id)
{
	const struct id_name *p = get_by_id(id.raw_id, product_table);
	return p ? p->name : NULL;
}

const char* unipi_product_dt(platform_id_t id)
{
	const struct id_name *p = get_by_id(id.raw_id, product_table);
	return p ? p->dt : NULL;
}

const char* unipi_board_name(board_model_t id)
{
	const struct id_name *p = get_by_id(id, board_table);
	return p ? p->name : NULL;
}

const char* unipi_board_dt(board_model_t id)
{
	const struct id_name *p = get_by_id(id, board_table);
	return p ? p->dt : NULL;
}

const char* unipi_slot_dt(board_model_t id, int slot)
{
	const struct id_name *p = get_by_id(id, board_table);
	const struct slot_dt *d;
	if (p == NULL )
		return NULL;
	d = get_by_slot(slot, p->slots);
	return d ? d->dt : NULL;
}

File: template/unipi-values.template.py
##########################################################################
#
#  {{ warning }}
#
#  uniee_values.py
#
#  Created on: Jan 14, 2022
#      Author: Miroslav Ondra <ondra@faster.cz>
# 
{% include "sss" ignore missing %}

class Product:
	def __init__(self, id, name, **kwargs):
		self.id = id
		self.name = name
		self.vars = kwargs

class Board:
	def __init__(self, id, name, slots, **kwargs):
		self.id = id
		self.name = name
		self.slots = slots
		self.vars = kwargs

class Slot:
	def __init__(self, slot, **kwargs):
		self.slot = slot
		self.vars = kwargs

products = {
{%- for pname,pid in platform.id.items() %}
  '{{ pid|int }}': Product({{ pid|int }}, '{{ pname }}'
    {%- if pid|trim in product_dt -%}
             {%- for k,v in product_dt[pid|trim].items() -%}, {{ k }}='{{ v }}' {% endfor -%}
    {%- endif -%}
    ),
{%- endfor %}
}

boards = {
{%- for bname,b_id in board.model.items() %}
  '{{ b_id|int }}': Board({{ b_id|int }}, '{{ bname }}'
    {%- if b_id|trim in board_dt -%}
    ,{
        {%- for s,sv in board_dt[b_id|trim].items() %}
             '{{s|int}}': Slot({{s|int}}
             {%- for k,v in sv.items() -%}, {{ k }}='{{ v }}' {% endfor -%}),
        {%- endfor %}
    }
    {#- %- if b_id|trim in board_dt -%}
    , dt="{{ board_dt[v|trim] }}"
    {%- endif -%  -#}
    {%- else -%}
    , None
    {%- endif -%}
           ),
{%- endfor %}
}

# Family names
family = {
{%- for f,v in platform.family.items() %}
  '{{ v|int }}': '{{ f }}',
{%- endfor %}
}

def unipi_product_info(product_id, version=None):
	''' return product_info or none '''
	return products.get(str(product_id), None)

def unipi_product_info_by_name(product_name, version=None):
	''' return product_info or none '''
	for k,v in products.items():
		if v.name.lower() == product_name.lower(): 
			return v
	return None

def unipi_board_info(board_id, slot=None, version=None):
	''' return board_info or None '''
	board_info = boards.get(str(board_id), None)
	if slot == None or board_info == None:
		return board_info
	if board_info.slots is None:
		return None
	return board_info.slots.get(str(slot), None)


def unipi_family_name(product_id):
	''' return family name or none '''
	return family.get(str(product_id & 0xff), None)

